src/Database/LSMTree/Internal/Snapshot.hs

```diff
@@ -166,6 +179,221 @@ instance (NFData t, NFData r) => NFData (SnapMergingRunState t r) where
+    fromSnapTreeState snapTreeState
+  where
+    mkPath = RunFsPaths (getActiveDir dir)
+
+    dupRun :: RefCounted m obj => Ref obj -> m (Ref obj)
+    dupRun r = withRollback reg (dupRef r) releaseRef
+
+    -- Used for inductive calls
+    recurrance :: SnapMergingTree (Ref (Run m h)) -> m (Ref (MT.MergingTree m h))
+    recurrance = fromSnapMergingTree reg hfs hbio conf uc resolve dir
+
+    fromSnapTreeState :: SnapMergingTreeState (Ref (Run m h)) -> m (Ref (MT.MergingTree m h))
+    fromSnapTreeState (SnapCompletedTreeMerge run) = dupRun run >>= \run' ->
+      MT.newPendingLevelMerge [MT.PreExistingRun run'] Nothing
+    fromSnapTreeState (SnapPendingTreeMerge pMerge) = case pMerge of
+      SnapPendingLevelMerge peRuns maybeMergeTree -> do
+        peRuns' <- traverse fromSnapPreExistingRun peRuns
+        maybeMergeTree' <- traverse recurrance maybeMergeTree
+        MT.newPendingLevelMerge peRuns' maybeMergeTree'
+      SnapPendingUnionMerge mergeTrees -> MT.newPendingUnionMerge =<< traverse recurrance mergeTrees
+    fromSnapTreeState (SnapOngoingTreeMerge nr ne (SuppliedCredits sc) smrs) = do
+        ongoingState <- getSnapMergingRunState nr ne sc smrs
+        mergeState <- newMVar $! MT.OngoingTreeMerge ongoingState
+        newRef (MT.finalise mergeState) $ \mergeRefCounter ->
+          MT.MergingTree { mergeState, mergeRefCounter }
+
+    getSnapMergingRunState
+      :: forall t.
+         MR.IsMergeType t
+      => NumRuns
+      -> NumEntries
+      -> Int
+      -> SnapMergingRunState t (Ref (Run m h))
+      -> m (Ref (MR.MergingRun t m h))
+    getSnapMergingRunState nr ne sc smrs = do
+          let ln = LevelNo 0
+              caching = diskCachePolicyForLevel confDiskCachePolicy ln
+              alloc = bloomFilterAllocForLevel conf ln
+              indexType = indexTypeForRun confFencePointerIndex
+          case smrs of
+            SnapCompletedMerge run ->
+              withRollback reg (MR.newCompleted nr ne run) releaseRef
+            SnapOngoingMerge runs mt -> do
+              rn <- uniqueToRunNumber <$> incrUniqCounter uc
+              mr <- withRollback reg
+                (MR.new hfs hbio resolve caching alloc indexType mt (mkPath rn) runs)
+                releaseRef
+              -- When a snapshot is created, merge progress is lost, so we
+              -- have to redo merging work here. SuppliedCredits tracks how
+              -- many credits were supplied before the snapshot was taken.
+              leftoverCredits <- MR.supplyCredits
+                                   mr (creditThresholdForLevel conf ln)
+                                   (MR.Credits sc)
+              assert (leftoverCredits == 0) $ return mr
+
+    fromSnapPreExistingRun :: SnapPreExistingRun (Ref (Run m h)) -> m (MT.PreExistingRun m h)
+    fromSnapPreExistingRun (SnapPreExistingRun run) = MT.PreExistingRun <$> dupRun run
```

Comment:
This `dupRun` is also fishy. We should try to follow our standard pattern of callee makes dups when they need to retain a reference, rather than caller. PR #588 goes further in this direction, but for incoming runs not merge trees. We should review this for merge trees.